File: python/external-user-service/OrderService/models/entities.py
"""
Database entities for Order Service
Equivalent to the .NET OrderService.Models.Entities
"""

from datetime import datetime
from typing import List, Optional
from sqlalchemy import Column, Integer, String, DateTime, Boolean, ForeignKey, Text
from sqlalchemy.types import DECIMAL
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import relationship

Base = declarative_base()


class Customer(Base):
    """Customer entity"""
    __tablename__ = "customers"
    
    id = Column(Integer, primary_key=True, index=True)
    first_name = Column(String(100), nullable=False)
    last_name = Column(String(100), nullable=False)
    email = Column(String(255), nullable=False, unique=True, index=True)
    phone = Column(String(20), nullable=True)
    address = Column(String(500), nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    
    # Relationships
    orders = relationship("Order", back_populates="customer", cascade="all, delete-orphan")


class Product(Base):
    """Product entity"""
    __tablename__ = "products"
    
    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(200), nullable=False)
    description = Column(String(1000), nullable=True)
    price = Column(DECIMAL(18, 2), nullable=False)
    sku = Column(String(50), nullable=True, unique=True, index=True)
    stock_quantity = Column(Integer, default=0, nullable=False)
    category = Column(String(100), nullable=True)
    is_active = Column(Boolean, default=True, nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    
    # Relationships
    order_items = relationship("OrderItem", back_populates="product")


class Order(Base):
    """Order entity"""
    __tablename__ = "orders"
    
    id = Column(Integer, primary_key=True, index=True)
    customer_id = Column(Integer, ForeignKey("customers.id"), nullable=False)
    order_number = Column(String(50), nullable=False, unique=True, index=True)
    order_date = Column(DateTime, default=datetime.utcnow, nullable=False)
    status = Column(String(50), default="Pending", nullable=False)  # Pending, Processing, Shipped, Delivered, Cancelled
    sub_total = Column(DECIMAL(18, 2), default=0, nullable=False)
    tax_amount = Column(DECIMAL(18, 2), default=0, nullable=False)
    shipping_amount = Column(DECIMAL(18, 2), default=0, nullable=False)
    total_amount = Column(DECIMAL(18, 2), default=0, nullable=False)
    notes = Column(String(500), nullable=True)
    shipping_address = Column(String(200), nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    
    # Relationships
    customer = relationship("Customer", back_populates="orders")
    order_items = relationship("OrderItem", back_populates="order", cascade="all, delete-orphan")


class OrderItem(Base):
    """Order item entity"""
    __tablename__ = "order_items"
    
    id = Column(Integer, primary_key=True, index=True)
    order_id = Column(Integer, ForeignKey("orders.id"), nullable=False)
    product_id = Column(Integer, ForeignKey("products.id"), nullable=False)
    quantity = Column(Integer, nullable=False)
    unit_price = Column(DECIMAL(18, 2), nullable=False)
    total_price = Column(DECIMAL(18, 2), nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    
    # Relationships
    order = relationship("Order", back_populates="order_items")
    product = relationship("Product", back_populates="order_items")

File: nodejs/external-user-service/OrderService/src/models/entities.ts
/**
 * Database entities for Order Service
 * TypeScript equivalent to the Python OrderService.Models.Entities
 */

import { Table, Column, Model, DataType, PrimaryKey, AutoIncrement, CreatedAt, UpdatedAt, HasMany, BelongsTo, ForeignKey } from 'sequelize-typescript';

@Table({
    tableName: 'customers',
    timestamps: true,
    underscored: true
})
export class Customer extends Model<Customer> {

    @PrimaryKey
    @Column({
        type: DataType.UUID,
        defaultValue: DataType.UUIDV4
    })
    id!: string;

    @Column({
        type: DataType.STRING(100),
        allowNull: false
    })
    first_name!: string;

    @Column({
        type: DataType.STRING(100),
        allowNull: false
    })
    last_name!: string;

    @Column({
        type: DataType.STRING(255),
        allowNull: false,
        unique: true
    })
    email!: string;

    @Column({
        type: DataType.STRING(20),
        allowNull: true
    })
    phone?: string;

    @Column({
        type: DataType.STRING(500),
        allowNull: true
    })
    address?: string;

    @CreatedAt
    @Column({
        type: DataType.DATE,
        allowNull: false
    })
    created_at!: Date;

    @UpdatedAt
    @Column({
        type: DataType.DATE,
        allowNull: false
    })
    updated_at!: Date;

    @HasMany(() => Order)
    orders!: Order[];
}

@Table({
    tableName: 'products',
    timestamps: true,
    underscored: true
})
export class Product extends Model<Product> {

    @PrimaryKey
    @AutoIncrement
    @Column({
        type: DataType.INTEGER
    })
    id!: number;

    @Column({
        type: DataType.STRING(200),
        allowNull: false
    })
    name!: string;

    @Column({
        type: DataType.STRING(1000),
        allowNull: true
    })
    description?: string;

    @Column({
        type: DataType.DECIMAL(18, 2),
        allowNull: false
    })
    price!: number;

    @Column({
        type: DataType.STRING(50),
        allowNull: true,
        unique: true
    })
    sku?: string;

    @Column({
        type: DataType.INTEGER,
        allowNull: false,
        defaultValue: 0
    })
    stock_quantity!: number;

    @Column({
        type: DataType.STRING(100),
        allowNull: true
    })
    category?: string;

    @Column({
        type: DataType.BOOLEAN,
        allowNull: false,
        defaultValue: true
    })
    is_active!: boolean;

    @CreatedAt
    @Column({
        type: DataType.DATE,
        allowNull: false
    })
    created_at!: Date;

    @UpdatedAt
    @Column({
        type: DataType.DATE,
        allowNull: false
    })
    updated_at!: Date;

    @HasMany(() => OrderItem)
    order_items!: OrderItem[];
}

@Table({
    tableName: 'orders',
    timestamps: true,
    underscored: true
})
export class Order extends Model<Order> {

    @PrimaryKey
    @AutoIncrement
    @Column({
        type: DataType.INTEGER
    })
    id!: number;

    @ForeignKey(() => Customer)
    @Column({
        type: DataType.UUID,
        allowNull: false
    })
    customer_id!: string;

    @Column({
        type: DataType.STRING(50),
        allowNull: false,
        unique: true
    })
    order_number!: string;

    @Column({
        type: DataType.DATE,
        allowNull: false,
        defaultValue: DataType.NOW
    })
    order_date!: Date;

    @Column({
        type: DataType.STRING(50),
        allowNull: false,
        defaultValue: 'Pending'
    })
    status!: string; // Pending, Processing, Shipped, Delivered, Cancelled

    @Column({
        type: DataType.DECIMAL(18, 2),
        allowNull: false,
        defaultValue: 0
    })
    sub_total!: number;

    @Column({
        type: DataType.DECIMAL(18, 2),
        allowNull: false,
        defaultValue: 0
    })
    tax_amount!: number;

    @Column({
        type: DataType.DECIMAL(18, 2),
        allowNull: false,
        defaultValue: 0
    })
    shipping_amount!: number;

    @Column({
        type: DataType.DECIMAL(18, 2),
        allowNull: false,
        defaultValue: 0
    })
    total_amount!: number;

    @Column({
        type: DataType.STRING(500),
        allowNull: true
    })
    notes?: string;

    @Column({
        type: DataType.STRING(200),
        allowNull: true
    })
    shipping_address?: string;

    @CreatedAt
    @Column({
        type: DataType.DATE,
        allowNull: false
    })
    created_at!: Date;

    @UpdatedAt
    @Column({
        type: DataType.DATE,
        allowNull: false
    })
    updated_at!: Date;

    @BelongsTo(() => Customer)
    customer!: Customer;

    @HasMany(() => OrderItem)
    order_items!: OrderItem[];
}

@Table({
    tableName: 'order_items',
    timestamps: true,
    underscored: true
})
export class OrderItem extends Model<OrderItem> {

    @PrimaryKey
    @AutoIncrement
    @Column({
        type: DataType.INTEGER
    })
    id!: number;

    @ForeignKey(() => Order)
    @Column({
        type: DataType.INTEGER,
        allowNull: false
    })
    order_id!: number;

    @ForeignKey(() => Product)
    @Column({
        type: DataType.INTEGER,
        allowNull: false
    })
    product_id!: number;

    @Column({
        type: DataType.INTEGER,
        allowNull: false
    })
    quantity!: number;

    @Column({
        type: DataType.DECIMAL(18, 2),
        allowNull: false
    })
    unit_price!: number;

    @Column({
        type: DataType.DECIMAL(18, 2),
        allowNull: false
    })
    total_price!: number;

    @CreatedAt
    @Column({
        type: DataType.DATE,
        allowNull: false
    })
    created_at!: Date;

    @UpdatedAt
    @Column({
        type: DataType.DATE,
        allowNull: false
    })
    updated_at!: Date;

    @BelongsTo(() => Order)
    order!: Order;

    @BelongsTo(() => Product)
    product!: Product;
}
